### PR TITLE
Abort control flow upon a `return`, `RET`, or `RVRT`.

### LINKED ISSUE #308: Abort control flow in control flow analysis/dead code analysis when `std::chain::revert()` is called
Similar to rust's `panic!()`

### PR DESCRIPTION
**Pull request by sezna**
This should unblock the final _hard_ blocker for swayswap.

## What does this do?
In Rust, if a branch aborts control flow, then its return type doesn't matter. [Here is an example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=945737317eb037180f94762f751b68b2).

We want this behavior so we can implement `.unwrap()` -- panicking if there is no value present in an `Option` requires that we don't type check the `else` branch in case of a deterministic abort.

## How does this do?
This PR introduces a `.deterministically_aborts()` method on AST nodes, expressions, and code blocks which returns a true if, in its control flow, there is a `RVRT`, `RET`, or `return` that is impossible to avoid. We then special case some type checking in branched code around that, as well as take it into consideration when type checking expressions against their annotations.

## Other things
While implementing this, I found a bug in `return` statement type checking -- we were not type checking returns that are on the RHS of a variable declaration. For example, the below compiles when it should not.
```rust
fn main() -> u64 {
  let x = return true;
}
```
This PR includes a fix for that.

Closes #308

### CODE REVIEW

File: test/src/e2e_vm_tests/test_programs/abort_control_flow/Forc.toml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[project]
+authors = ["Fuel Labs <contact@fuel.sh>"]
+entry = "main.sw"
+license = "Apache-2.0"
+name = "abort_control_flow"
+
+[dependencies]
+core = { git = "http://github.com/FuelLabs/sway-lib-core" }
+std = { git = "http://github.com/FuelLabs/sway-lib-std" }
```
Comment (adlerjohn) [inline test/src/e2e_vm_tests/test_programs/abort_control_flow/Forc.toml:9]: Could you add a branch for these, and the manifest below? Also, could remove unused dependencies (I think core is unused).

**Comment by otrho** [comment]
LGTM otherwise.

**Review by adlerjohn — COMMENTED** [review]
Merge conflicts. Could you move the testcase into the directory structure of #965?

I've tested locally and can confirm that old-compiler does not compile the testcase and new-compiler does.

File: test/src/e2e_vm_tests/test_programs/should_fail/abort_control_flow/Forc.toml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+[project]
+authors = ["Fuel Labs <contact@fuel.sh>"]
+entry = "main.sw"
+license = "Apache-2.0"
+name = "abort_control_flow"
+
+[dependencies]
+std = { git = "http://github.com/FuelLabs/sway-lib-std" }
```
Comment (adlerjohn) [inline test/src/e2e_vm_tests/test_programs/should_fail/abort_control_flow/Forc.toml:8]: Branch here too, also change them all to `https`

**Review by adlerjohn — COMMENTED** [review]
LGTM, will defer to the actual compiler devs.

**Review by otrho — APPROVED** [review]

**Pull request merged**